README.md
# Summary
A basic RESTful API framework written in PHP. Allows you to create a web service where data can be accessed in a simple URL format such as **http://example.com/api/[resource]/[id]**.

# Installation
Simply deploy the files on your web server. Then, create your controllers as in the **classes/controllers/** directory. The sample `NewsController` uses a flat-file approach to storing data, but this can be easily extended to use a database or other data source.

### I have used this API in several projects and it is my favorite **drop-in** PHP API.

Just create a controller or a model and it's in!


data/news.txt
a:2:{i:1;a:4:{s:2:"id";i:1;s:5:"title";s:12:"Test Article";s:7:"content";s:34:"Welcome to your new API framework!";s:9:"published";s:25:"2012-01-13T18:35:48+00:00";}i:2;a:4:{s:2:"id";i:2;s:5:"title";s:11:"New Article";s:7:"content";s:62:"This is a great piece about how the world is unsafe and stuff.";s:9:"published";s:25:"2017-03-09T02:38:58+00:00";}}

index.php
<?php

/*
ANDY:
Some headers that allow cross domain requests, credentials, and a good set of API verbs.
*/
header("Access-Control-Allow-Origin: " . $_SERVER["HTTP_ORIGIN"]);
header("Access-Control-Allow-Headers: X-Requested-With, X-Authorization, Content-Type, X-HTTP-Method-Override");
header("Access-Control-Allow-Credentials: true");
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS, PATCH');

//ANDY: I have my own exception for deep in the API
require_once(dirname(__FILE__) . '/classes/ControllerException.php');

/**
 * API framework front controller.
 * 
 * @package api-framework
 * @author  Martin Bean <martin@martinbean.co.uk>
 */

/**
 * Generic class autoloader.
 * 
 * @param string $class_name
 */
function autoload_class($class_name) {
    $directories = array(
        'classes/',
        'classes/controllers/',
        'classes/models/'
    );
    foreach ($directories as $directory) {
        $filename = $directory . $class_name . '.php';
        if (is_file($filename)) {
            require($filename);
            break;
        }
    }
}

/**
 * Register autoloader functions.
 */
spl_autoload_register('autoload_class');

/**
 * Parse the incoming request.
 */
$request = new Request();
if (isset($_SERVER['PATH_INFO'])) {
    $request->url_elements = explode('/', trim($_SERVER['PATH_INFO'], '/'));
}

//ANDY:  When doing credentials you get an OPTIONS request first
$request->method = isset($_SERVER['REQUEST_METHOD'])? strtoupper($_SERVER['REQUEST_METHOD']) : 'OPTIONS';
switch ($request->method) {
    case 'GET':
        $request->parameters = $_GET;
    break;
    case 'POST': //ANDY: More elaborate POST so I can handle JSON inputs, form data, and standard POSTs
        $body = file_get_contents("php://input");
        $content_type = false;
        if(isset($_SERVER['CONTENT_TYPE'])) {
            $content_type = $_SERVER['CONTENT_TYPE'];
        }
        switch($content_type){
            case "application/json":
                $body_params = json_decode($body);
                if($body_params) {
                    foreach($body_params as $param_name => $param_value) {
                        $request->parameters[$param_name] = $param_value;
                    }
                }
                break;
            case "application/x-www-form-urlencoded":
                parse_str($body, $postvars);
                foreach($postvars as $field => $value) {
                    $request->parameters[$field] = $value;
                }
                break;
            default:
                $request->parameters = $_POST;
                break;
        }
    case 'PUT':
        parse_str(file_get_contents('php://input'), $request->parameters);
    break;
    case 'OPTIONS':
        //no-op
        return;
    default:
        $request->parameters = json_decode(file_get_contents('php://input'));
}

/**
 * Route the request.
 */
if (!empty($request->url_elements)) {
    $controller_name = ucfirst($request->url_elements[0]) . 'Controller';
    if (class_exists($controller_name)) {
        $controller = new $controller_name;
        $action_name = strtolower($request->method);
        $response_str = call_user_func_array(array($controller, $action_name), array($request));
    }
    else {
        header('HTTP/1.1 404 Not Found');
        $response_str = 'Unknown request: ' . $request->url_elements[0];
    }
}
else {
    $response_str = 'Unknown request';
}

/**
 * Send the response to the client.
 */
$response_obj = Response::create($response_str, $_SERVER['HTTP_ACCEPT']);
echo $response_obj->render();


classes/controllers/NewsController.php
<?php
/**
 * Sample news controller.
 * 
 * @package api-framework
 * @author  Martin Bean <martin@martinbean.co.uk>
 */
class NewsController extends AbstractController
{
    /**
     * News file.
     *
     * @var variable type
     */
    protected $articles_file = './data/news.txt';
    
    /**
     * GET method.
     * 
     * @param  Request $request
     * @return string
     */
    public function get($request)
    {
        $articles = $this->readArticles();
        switch (count($request->url_elements)) {
            case 1:
                return $articles;
            break;
            case 2:
                $article_id = $request->url_elements[1];
                return $articles[$article_id];
            break;
        }
    }
    
    /**
     * POST action.
     *
     * @param  $request
     * @return null
     */
    public function post($request)
    {
        switch (count($request->url_elements)) {
            case 1:
                $articles = unserialize(file_get_contents($this->articles_file));
                $id = (count($articles) + 1);
                $articles = $this->readArticles();
                $article = array(
                    'id' => $id,
                    'title' => $request->parameters['title'],
                    'content' => $request->parameters['content'],
                    'published' => date('c')
                );
                $articles[] = $article;
                $this->writeArticles($articles);
                header('HTTP/1.1 201 Created');
                header('Location: /news/'.$id);
                return null;
            break;
        }
    }
    
    /**
     * Read articles.
     *
     * @return array
     */
    protected function readArticles()
    {
        $articles = unserialize(file_get_contents($this->articles_file));
        if (empty($articles)) {
            $articles = array(
                1 => array(
                    'id' => 1,
                    'title' => 'Test Article',
                    'content' => 'Welcome to your new API framework!',
                    'published' => date('c', mktime(18, 35, 48, 1, 13, 2012))
                )
            );
            $this->writeArticles($articles);
        }
        return $articles;
    }
    
    /**
     * Write articles.
     *
     * @param  string $articles
     * @return boolean
     */
    protected function writeArticles($articles)
    {
        file_put_contents($this->articles_file, serialize($articles));
        return true;
    }
}